### PR TITLE
Fix Substrate features

### PR DESCRIPTION
**Pull request by ggwpez**
Trying to automatically fix `std`, `runtime-benchmarks` and `try-runtime` features with
```bash
zepter lint propagate-feature --feature try-runtime --left-side-feature-missing=ignore --workspace --fix --feature-enables-dep="try-runtime:frame-try-runtime"
zepter lint propagate-feature --feature runtime-benchmarks --left-side-feature-missing=ignore --workspace --fix --feature-enables-dep="runtime-benchmarks:frame-benchmarking"
zepter lint propagate-feature --feature std --left-side-feature-missing=ignore --workspace --fix
```
The auto-fixer tries to be as non-invasive as possible, but removes empty lines.

Explanation:
- This adds the feature for all dependencies that expose it. It does not create the feature if it is not present, since `left-side-feature-missing` is passed. To be really correct I think we would have to add it to all crates that do not have it though.
- The `feature-enables-dep` will enable the given dependency as non-optional when the feature is enabled. Otherwise the features of optional dependencies will also be enabled as optional.

Known blindspot: The `--workspace` flag currently requires that both sides are in the workspace. This will be extended soon to also allow the right-hand-side dependency to be outside of the worspace. Since currently it does not detect `serde/std` as missing when the `--workspace` flag is present.

TODO:
- [ ] Double check that it also works fine for dev-dependencies. I am not sure what rules to use here.
- [x] Add CI
- [ ] Add other [`featalign`](https://github.com/hack-ink/cargo-featalign) to double-check (will maybe do as follow up so we already have something working).

### CODE REVIEW

**Review by alvicsam — APPROVED** [review]

**Review by davxy — APPROVED** [review]

**Review by liamaharon — APPROVED** [review]

**Review by shawntabrizi — APPROVED** [review]

**Comment by shawntabrizi** [comment]
Could we use Zepter to a little bit of toml formatting too?

For example organizing dependencies alphabetically?

**Comment by ggwpez** [comment]
> Could we use Zepter to a little bit of toml formatting too?
> 
> For example organizing dependencies alphabetically?

Yes we definitely can. Currently the autofixer is written to not re-order anything, but it already has to touch the toml array, so it could be ordered as well.  
Comments within the `dependency` array are lost, but i think that is a fine trade-off in this case.

**Comment by koute** [comment]
> Comments within the `dependency` array are lost, but i think that is a fine trade-off in this case.

Could we make it so that they're not lost? (I assume there are `toml` libraries which can do this?)

**Comment by ggwpez** [comment]
> > Comments within the `dependency` array are lost, but i think that is a fine trade-off in this case.
> 
> Could we make it so that they're not lost? (I assume there are `toml` libraries which can do this?)

Oh, the official [`toml_edit`](https://docs.rs/toml_edit/latest/toml_edit/) crate actually mentions that it can preserve comment - so I must just be using it wrong. I will look into it, thanks!

**Review by sergejparity — APPROVED** [review]

**Review by sam0x17 — APPROVED** [review]

File: primitives/consensus/babe/Cargo.toml [hunk 1]
```diff
@@ -41,6 +41,7 @@ std = [
 	"sp-runtime/std",
 	"sp-std/std",
 	"sp-timestamp",
+	"sp-timestamp?/std"
```
Comment (bkchr) [inline primitives/consensus/babe/Cargo.toml:44]:
```suggestion
 	"sip-timestamp/std",
```

File: primitives/consensus/grandpa/Cargo.toml [hunk 2]
```diff
@@ -40,6 +40,7 @@ std = [
 	"sp-keystore",
 	"sp-runtime/std",
 	"sp-std/std",
+	"sp-keystore?/std"
```
Comment (bkchr) [inline primitives/consensus/grandpa/Cargo.toml:43]:
```suggestion
 	"sp-keystore/std",
 	"sp-runtime/std",
 	"sp-std/std"
```

File: primitives/core/Cargo.toml [hunk 3]
```diff
@@ -114,6 +114,7 @@ std = [
 	"libsecp256k1/std",
 	"dyn-clonable",
 	"tracing",
+	"sp-externalities?/std"
```
Comment (bkchr) [inline primitives/core/Cargo.toml:117]: Std feature enabling should be moved up to the enabling of the crate

File: primitives/io/Cargo.toml [hunk 4]
```diff
@@ -60,6 +60,7 @@ std = [
 	"log",
 	"ed25519-dalek",
 	"ed25519",
+	"sp-keystore?/std"
```
Comment (bkchr) [inline primitives/io/Cargo.toml:63]: Same

File: primitives/session/Cargo.toml [hunk 5]
```diff
@@ -33,4 +33,5 @@ std = [
 	"sp-staking/std",
 	"sp-std/std",
 	"sp-keystore",
+	"sp-keystore?/std"
```
Comment (bkchr) [inline primitives/session/Cargo.toml:36]:
```suggestion
 	"sp-keystore/std"
```

File: primitives/transaction-storage-proof/Cargo.toml [hunk 6]
```diff
@@ -33,4 +33,5 @@ std = [
 	"sp-runtime/std",
 	"sp-std/std",
 	"sp-trie/std",
+	"sp-core?/std"
```
Comment (bkchr) [inline primitives/transaction-storage-proof/Cargo.toml:36]: Same

File: frame/asset-rate/Cargo.toml [hunk 7]
```diff
@@ -40,16 +40,20 @@ std = [
...
 	"sp-core?/std",
+	"pallet-balances/std",
+	"sp-io/std"
 ]
 runtime-benchmarks = [
 	"frame-benchmarking/runtime-benchmarks",
 	"frame-support/runtime-benchmarks",
 	"frame-system/runtime-benchmarks",
 	"sp-runtime/runtime-benchmarks",
 	"dep:sp-core",
```
Comment (bkchr) [inline frame/asset-rate/Cargo.toml:51]:
```suggestion
 	"sp-core",
```

File: frame/examples/offchain-worker/Cargo.toml [hunk 8]
```diff
@@ -39,5 +39,10 @@ std = [
 	"sp-keystore",
 	"sp-runtime/std",
 	"sp-std/std",
+	"sp-keystore?/std"
```
Comment (bkchr) [inline frame/examples/offchain-worker/Cargo.toml:42]:
```suggestion
 	"sp-keystore/std",
 	"sp-runtime/std",
 	"sp-std/std"
```

File: frame/support/test/Cargo.toml [hunk 9]
```diff
@@ -53,12 +53,14 @@ std = [
 	"sp-std/std",
 	"sp-version/std",
 	"test-pallet/std",
+	"sp-state-machine?/std"
```
Comment (bkchr) [inline frame/support/test/Cargo.toml:56]: Same

**Review by bkchr — APPROVED** [review]
Looks good besides the edge case around enabling of a crate and a feature in one go that I highlighted in the review

**Comment by ggwpez** [comment]
bot merge

**Comment by AurevoirXavier** [comment]
[cargo-featalign](https://github.com/hack-ink/cargo-featalign) can address many of the situations mentioned in this thread such as preseving comments. Recently, I added support for sorting features alphabetically based on @shawntabrizi's comment.

However, I am unsure about how parity CI functions. I am eager to apply this tool to Substrate.

**Comment by ggwpez** [comment]
@AurevoirXavier you can check out the diff in [`scripts/ci/gitlab/pipeline/check.yml`](https://github.com/paritytech/substrate/pull/14660/files#diff-b48ad7f87648ee915e6ba8cea75cf43f61a915f35491bdb2097ae8be9ab90415) of this MR.  
You can just add featalign alongside the Zepter invocation there into the same job. I am not a CI engie myself, but it should be fine.  
The [job](https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/3311804) currently takes 33 sec, it currently runs as one of the first jobs, but AFAIK it does not block later ones.
